websockify/websocket.py

```diff
@@ -1,1109 +1,825 @@
-            try:
-                f['payload'] = b64decode(f['payload'])
-            except:
-                logger.exception("Exception while b64decoding buffer: %s" %
-                                 (repr(buf)))
-                raise
+            self._state = "send_headers"
 
-        if f['opcode'] == 0x08:
-            if f['length'] >= 2:
-                f['close_code'] = unpack_from(">H", f['payload'])[0]
-            if f['length'] > 3:
-                f['close_reason'] = f['payload'][2:]
+        if self._state == "send_headers":
+            self._flush()
+            self._state = "response"
 
-        return f
+        if self._state == "response":
+            if not self._recv():
+                raise Exception("Socket closed unexpectedly")
 
+            if self._recv_buffer.find('\r\n\r\n'.encode("ascii")) == -1:
+                raise WebSocketWantReadError
 
-    #
-    # WebSocketRequestHandler logging/output functions
-    #
-
-    def print_traffic(self, token="."):
-        """ Show traffic flow mode. """
-        if self.traffic:
-            sys.stdout.write(token)
-            sys.stdout.flush()
-
-    def msg(self, msg, *args, **kwargs):
-        """ Output message with handler_id prefix. """
-        prefix = "% 3d: " % self.handler_id
-        self.logger.log(logging.INFO, "%s%s" % (prefix, msg), *args, **kwargs)
-
-    def vmsg(self, msg, *args, **kwargs):
-        """ Same as msg() but as debug. """
-        prefix = "% 3d: " % self.handler_id
-        self.logger.log(logging.DEBUG, "%s%s" % (prefix, msg), *args, **kwargs)
-
-    def warn(self, msg, *args, **kwargs):
-        """ Same as msg() but as warning. """
-        prefix = "% 3d: " % self.handler_id
-        self.logger.log(logging.WARN, "%s%s" % (prefix, msg), *args, **kwargs)
-
-    #
-    # Main WebSocketRequestHandler methods
-    #
-    def send_frames(self, bufs=None):
-        """ Encode and send WebSocket frames. Any frames already
-        queued will be sent first. If buf is not set then only queued
-        frames will be sent. Returns the number of pending frames that
-        could not be fully sent. If returned pending frames is greater
-        than 0, then the caller should call again when the socket is
-        ready. """
-
-        tdelta = int(time.time()*1000) - self.start_time
-
-        if bufs:
-            for buf in bufs:
-                if self.base64:
-                    encbuf, lenhead, lentail = self.encode_hybi(buf, opcode=1, base64=True)
-                else:
-                    encbuf, lenhead, lentail = self.encode_hybi(buf, opcode=2, base64=False)
+            (request, self._recv_buffer) = self._recv_buffer.split('\r\n'.encode("ascii"), 1)
+            request = request.decode("latin-1")
 
-                if self.rec:
-                    self.rec.write("%s,\n" %
-                            repr("{%s{" % tdelta
-                                + encbuf[lenhead:len(encbuf)-lentail]))
+            headers = {}
```

Comment:
Probably. I seem to recall it was messy though and I didn't feel it was worth the effort. I can have another look.